dpar/src/features/input_layers.rs

```diff
@@ -299,21 +299,59 @@ impl InputVectorizer {
                     address: vec![addr.dependent],
                     layer: addr::Layer::Token,
                 };
+                let addr_head_pos = addr::AddressedValue {
+                    address: vec![addr.head],
+                    layer: addr::Layer::Tag,
+                };
+                let addr_dependent_pos = addr::AddressedValue {
+                    address: vec![addr.dependent],
+                    layer: addr::Layer::Tag,
+                };
+
                 let head = addr_head.get(state);
                 let dependent = addr_dependent.get(state);
-                if let (Some(head), Some(dependent)) = (head, dependent) {
-                    let association = self.assoc_strength(&head, &dependent, &deprel);
-                    non_lookup_slice[idx] = association;
+                let head_pos = addr_head_pos.get(state);
+                let dependent_pos = addr_dependent_pos.get(state);
+
+                if let (Some(head), Some(dependent), Some(head_pos), Some(dependent_pos)) =
+                    (head, dependent, head_pos, dependent_pos)
+                {
+                    let (association, found) =
+                        self.assoc_strength(&head, &dependent, &head_pos, &dependent_pos, &deprel);
+                    non_lookup_slice[idx * 2] = association;
+                    non_lookup_slice[idx * 2 + 1] = found;
                 }
             }
         }
     }
 
-    fn assoc_strength(&self, head: &str, dependent: &str, deprel: &str) -> f32 {
-        let dep_triple = (head.to_string(), dependent.to_string(), deprel.to_string());
+    /// Look up association measures in a `HashMap`.
+    ///
+    /// Return association measure and a boolean that indicates whether
+    /// the measure was retrieved for the dependency triple or if the
+    /// dependency triple did not exist and the default had to be picked.
+    fn assoc_strength(
+        &self,
+        head: &str,
+        dependent: &str,
+        head_pos: &str,
+        dependent_pos: &str,
+        deprel: &str,
+    ) -> (f32, f32) {
+        let mut head = head.to_string();
+        let mut dependent = dependent.to_string();
+
+        if head != "ROOT" && head_pos != "NN" && head_pos != "NE" {
+            head = head.to_lowercase();
+        }
+        if dependent != "ROOT" && dependent_pos != "NN" && dependent_pos != "NE" {
+            dependent = dependent.to_lowercase();
+        }
+
+        let dep_triple = (head, dependent, deprel.to_string());
         match self.association_strengths.get(&dep_triple) {
-            Some(association_strength) => association_strength.to_owned(),
-            None => 0.5.to_owned(),
+            Some(association_strength) => (association_strength.to_owned(), 1.0),
```

Comment:
f32 is a copy type. deref:

```
Some(association_strength) => (*association_strength, 1.0),
```

or pattern match:

```
Some(&association_strength) => (association_strength, 1.0),
```

General rules:

* Reference to copy types: deref or pattern match.
* Get owned value `T` from a reference `&T` where `T` is not a copy type: `.clone()`
* Get owned value `T` from a reference that is not `&T`: `.to_owned()` This is really a special case for special reference types such as `&str -> String` and `&[] -> Vec`.

What you do compiles, because there is a blanket implementation of `ToOwned` for types that implement `Clone` (and `Copy` requires `Clone`), but the reader would now expect that `association_strength` is something like a slice reference.

Why does Rust even provide these blanket implementations of `ToOwned` for `Clone` types? Good question! ;) This is to be able to write methods such as:

~~~
fn yadda(some_arg: impl ToOwned) {
  let owned = some_arg.to_owned();
  // ...
}
~~~

Which worksfor special reference types, clone types, and copy types.